Add "warnings", deprecate "warning" in {create,load,unload,restore}wallet

jonatack opened the pull request:
Based on discussion and concept ACKed in #27138, add a `warnings` field to RPCs createwallet, loadwallet, unloadwallet, and restorewallet as a JSON array of strings to replace the `warning` string field in these 4 RPCs. The idea is to more gracefully handle multiple warning messages and for consistency with other wallet RPCs.  Then, deprecate the latter fields, which represent all the remaining RPC `warning` fields.

The first commit https://github.com/bitcoin/bitcoin/pull/27279/commits/f73782a9032a462a71569e9424db9bf9eeababf3 implements https://github.com/bitcoin/bitcoin/pull/27138#issuecomment-1474789198 as an alternative to #27138. One of those two could potentially be backported to our currently supported releases.

jonatack committed:
rpc: extract wallet "warnings" fields to a util helper

jonatack committed:
rpc: add "warnings" field to RPCs {create,load,unload,restore}wallet

This new "warnings" field is a JSON array of strings intended to replace the
"warning" string field in these four RPCs, to better handle returning multiple
warning messages and for consistency with other wallet RPCs.

jonatack committed:
doc: fix/improve warning helps in {create,load,unload,restore}wallet

- clarify that there can be multiple warning messages
- specify the correct wallet action
- describe the use of newlines as delimiters

jonatack committed:
test: add test coverage for "warnings" field in createwallet

and clarify the "warning" field behavior.

jonatack committed:
rpc: deprecate "warning" field in {create,load,unload,restore}wallet

This string field has been replaced in these four RPCs by a "warnings" field
returning a JSON array of strings.

vasild commented on src/wallet/rpc/backup.cpp:
```diff
@@ -1932,7 +1938,13 @@ RPCHelpMan restorewallet()
     UniValue obj(UniValue::VOBJ);
     obj.pushKV("name", wallet->GetName());
-    obj.pushKV("warning", Join(warnings, Untranslated("\n")).original);
+    if (IsDeprecatedRPCEnabled("warningfield")) {
+        // The "warning" field is deprecated in v25 for removal in v26.
+        obj.pushKV("warning", Join(warnings, Untranslated("\n")).original);
+    }
+    if (!warnings.empty()) {
+        obj.pushKV("warnings", BilingualStringsToUniValue(warnings));
+    }
```
This is repeated in a few places and can be deduplicated (untested):

```cpp
void PushWarnings(const std::vector<bilingual_str>& warnings, UniValue& to)
{
    if (IsDeprecatedRPCEnabled("warningfield")) {
        // The "warning" field is deprecated in v25 for removal in v26.
        to.pushKV("warning", Join(warnings, Untranslated("\n")).original);
    }
    if (warnings.empty()) {
        return;
    }
    to.pushKV("warnings", UniValue{UniValue::VARR});
    for (const auto& s : warnings) {
        to["warnings"].push_back(s.original);
    }
}
```

vasild reviewed (commented):
Concept ACK

jonatack committed:
doc: release note for wallet RPCs "warning" field deprecation

jonatack committed:
rpc: move WALLET_FLAG_CAVEATS to the compilation unit of its caller

and add the walletutil.h include header for WALLET_FLAG_AVOID_REUSE that was
already missing before this change.

WALLET_FLAG_CAVEATS is only used in one RPC, so no need to encumber wallet.h and
wallet.cpp with it, along with all of the files that include wallet.h during
their compilation. Also apply clang-format per:

git diff -U0 HEAD~1.. | ./contrib/devtools/clang-format-diff.py -p1 -i -v

jonatack committed:
test: createwallet "warning" field deprecation test

jonatack committed:
test: fix importmulti/importdescriptors assertion

as these RPCs have a "warnings" field, not a "warning" one.

pinheadmz reviewed (commented):
concept ACK

Light code review. I prefer this to #27138 it seems like the right move. Built and ran tests, played with interface, nice!

```
--> bccli -regtest -named createwallet wallet_name=oops2 descriptors=false passphrase=""
{
  "name": "oops2",
  "warnings": [
    "Empty string given as passphrase, wallet will not be encrypted.",
    "Wallet created successfully. The legacy wallet type is being deprecated and support for creating and opening legacy wallets will be removed in the future."
  ]
}
```

I did a quick search for other RPCs that string together warnings, I think `getblockchaininfo` still does - worth including here?

Also, the GUI: I don't think encrypting wallet with a blank passphrase is possible from the UI so I don't know if there is a way to see multiple warnings in production, or if that is already handled.

pinheadmz commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -265,7 +265,12 @@ def run_test(self):
             load_res = node_master.loadwallet("u1_v16")
             # Make sure this wallet opens without warnings. See https://github.com/bitcoin/bitcoin/pull/19054
-            assert_equal(load_res['warning'], '')
+            if '"warnings"' in node_master.help("loadwallet"):
```
Is this the best way to test for this? Seems like it could be broken by someone using a word in a future doc update. You can't get `bitcoind --version` or something for the back-compat test?

pinheadmz commented on test/functional/wallet_createwallet.py:
```diff
@@ -25,6 +25,7 @@ def add_options(self, parser):
 
     def set_test_params(self):
         self.num_nodes = 1
+        self.extra_args = [["-deprecatedrpc=walletwarningfield"]]
```
Why do you set this flag for the entire test? Since the `warnings` is the new thing going forward, wouldn't it make more sense to write all the tests with only that field by default? And just add one test case with the flag to check `warning` was also included?

pinheadmz commented on src/wallet/wallet.h:
```diff
@@ -145,8 +145,6 @@ static const std::map<std::string,WalletFlags> WALLET_FLAG_MAP{
     {"external_signer", WALLET_FLAG_EXTERNAL_SIGNER}
 };
 
-extern const std::map<uint64_t,std::string> WALLET_FLAG_CAVEATS;
```
> so no need to encumber wallet.h and wallet.cpp with it

I'm not sure I totally agree with this, but just because all the flags are defined here already. Like, you'd need to import this header file to use `WALLET_FLAG_AVOID_REUSE` anyway, why not keep the caveats right next to their associated flags?

pinheadmz reviewed (commented):
code review ACK @ e67da5fedd526c38e030d6d18f3678313a683dc9

A few questions.

sedited reviewed (commented):
Approach ACK

vasild commented on src/wallet/rpc/backup.cpp:
```diff
@@ -1903,7 +1903,11 @@ RPCHelpMan restorewallet()
             {
                 {RPCResult::Type::STR, "name", "The wallet name if restored successfully."},
-                {RPCResult::Type::STR, "warning", "Warning message if wallet was not loaded cleanly."},
+                {RPCResult::Type::STR, "warning", /*optional=*/true, "Warning messages, if any, related to restoring the wallet. Multiple messages will be delimited by newline escape sequences (\"\\n\"). (DEPRECATED, returned only if config option -deprecatedrpc=walletwarningfield is passed.)"},
```
nit: actual newlines are used (`0A`), not newline escape sequences (`5C 6E`), thus:

```suggestion
                {RPCResult::Type::STR, "warning", /*optional=*/true, "Warning messages, if any, related to restoring the wallet. Multiple messages will be delimited by newlines. (DEPRECATED, returned only if config option -deprecatedrpc=walletwarningfield is passed.)"},
```

vasild commented on test/functional/wallet_createwallet.py:
```diff
@@ -174,8 +181,24 @@ def run_test(self):
+                "warnings": [EMPTY_PASSPHRASE_MSG, LEGACY_WALLET_MSG],
+            })
+
+        self.log.info('Test "warning" field deprecation, i.e. not returned without -deprecatedrpc=walletwarningfield')
+        self.restart_node(0, extra_args=[])
+        result = self.nodes[0].createwallet(wallet_name="w7_again", disable_private_keys=False, blank=False, passphrase="")
+        assert_equal(set([*result]), {"name", "warnings"})
```
nit: this would needlessly fail if a new unrelated field is added. Here we actually care that `warning` is not in the list, something like `assert("warning" not in result);`

vasild commented on src/wallet/rpc/backup.cpp:
```diff
@@ -1379,7 +1379,7 @@ RPCHelpMan importmulti()
         for (const UniValue& data : requests.getValues()) {
             const int64_t timestamp = std::max(GetImportTimestamp(data, now), minimumTimestamp);
-            const UniValue result = ProcessImport(*pwallet, data, timestamp);
+            const UniValue& result = ProcessImport(*pwallet, data, timestamp);
```
Commit 363b422983e01a532df67b2432cb13e47bf092e0 `rpc: rm unneeded UniValue copies in importmulti and importdescriptors` is useless. There is not any copying happening and the old and the new code are the same wrt how many objects are being created and destroyed.

```cpp
/* The rule of five. */
class R5
{
public:
    R5(int x, double y)
    {
        x_ = x;
        y_ = y;
        std::cout << "R5::R5() this=" << this << " x=" << x_ << ", y=" << y_ << std::endl;
    }

    ~R5()
    {
        std::cout << "R5::~R5() this=" << this << " x=" << x_ << ", y=" << y_ << std::endl;
    }

    R5(const R5& other)
#if 0
        = delete;
#else
    {
        std::cout << "R5::R5(const R5&) this=" << this << " other=" << &other << " x=" << other.x_
                  << ", y=" << other.y_ << std::endl;
        x_ = other.x_;
        y_ = other.y_;
    }
#endif

    R5(R5&& other)
    {
        std::cout << "R5::R5(R5&&) this=" << this << " other=" << &other << " x=" << other.x_
                  << ", y=" << other.y_ << std::endl;
        x_ = other.x_;
        y_ = other.y_;
        other.x_ = 0;
        other.y_ = 0;
    }

    R5& operator=(const R5& other)
#if 0
        = delete;
#else
    {
        std::cout << "R5::operator=(const R5&) this=" << this << " x=" << x_ << ", y=" << y_
                  << " other=" << &other << " x=" << other.x_ << ", y=" << other.y_ << std::endl;
        x_ = other.x_;
        y_ = other.y_;
        return *this;
    }
#endif

    R5& operator=(R5&& other)
    {
        std::cout << "R5::operator=(R5&&) this=" << this << " x=" << x_ << ", y=" << y_
                  << " other=" << &other << " x=" << other.x_ << ", y=" << other.y_ << std::endl;
        x_ = other.x_;
        y_ = other.y_;
        other.x_ = 0;
        other.y_ = 0;
        return *this;
    }

    std::ostream& operator<<(std::ostream& os) { return os << "(x=" << x_ << ", y=" << y_ << ")"; }

    int x_;
    double y_;
};

R5 f()
{
    R5 result{1, 2.3};
    result.x_ = 11;
    return result;
}

int main(int, char**)
{
    std::cout << "before f()\n";
    const R5 r5 = f();
    std::cout << "after f()\n";

    return 0;
}
```

prints the same with `const R5 r5` and with `const R5& r5`.

vasild replied:
My code is very similar to the example in https://en.cppreference.com/w/cpp/language/copy_elision. From that doc it seems compilers are not required to do that optimization, which was my thinking. Anyway, if all supported compilers do that optimization, then this change is still unnecessary.

vasild reviewed (approved):
ACK e67da5fedd526c38e030d6d18f3678313a683dc9

Some non-blockers below.

1440000bytes reviewed (approved):
utACK https://github.com/bitcoin/bitcoin/pull/27279/commits/7ccdd741fe1544c13b2a9b7baa5c5727e84d6e55

vasild reviewed (approved):
ACK 7ccdd741fe1544c13b2a9b7baa5c5727e84d6e55

pinheadmz reviewed (approved):
re-ACK 7ccdd741fe1544c13b2a9b7baa5c5727e84d6e55

confirmed trivial diff since last review, ran unit and functional tests locally, ran in regtest one more time to check UX with and without deprecated rpc flag

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 7ccdd741fe1544c13b2a9b7baa5c5727e84d6e55
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmQ20YEACgkQ5+KYS2KJ
yTqF/Q/8DnfN9Esae47Y1bvswkY5oX93SBB59pUEOHSxfTugRWFr18xZyUZI3zHz
kK8SFsVEXfm7lWTdMiRLjDSthFcsHMnzv+JNc76qvFuOhD5TYHkX81V7fUXSiDiS
VaODlGlVS/Cw6KoOQ52WWXltJxt0bIxPZMYBrtSus7poarSuCaOsXDqCaU+Ma6SM
KglES3j1+M4f7FIPUD6QVzHbhfYddTnFuwsBAgwMBxb/c7O8JrEWwxem80vc2jH4
9erbjBhF8YotvH4ici8Xl0bO6/FmpldLz3CNRqbl7tYJOIxU+z+WsgpcI+D9C2jQ
NXOIL6bKCHom8MlOBJpTez892wRoTjV6SXNHa1TrlGuRFUVQpTb8QWMiP4RmSuFm
UcDyuwLkDOUGx09SviS2FjZA1BBogCsWiwD7ZpoBcW1sXqmFqGL9VCGrAzmuwqcd
9xDn9vbzcvSEx8JdQvDYRpjdnQ8y2yt1VsgbpABRtb8uj16GLT1WiDCMFJFGxkV+
3dSIIHe7DnNSw30v8YFoBYCbrXQBJLhNGu1AsLmlQJ3rpiULOAklRqWV4tE22cp8
BU8g9003FTC5gBTJlyXDdZTVkh82ucuTWAlpFQysHXswunmP6nmYvMDhWjSPf46d
jdhQo38SR7lZQ49cetwAOOy2G7TxOm7T0/vts2HUCp8Pyx6sjBo=
=Q/2X
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

achow101 commented:
ACK 7ccdd741fe1544c13b2a9b7baa5c5727e84d6e55
